### ISSUE TITLE
New [336.v182c0fbaaeb7] release breaks SCM checkout

### ISSUE THREAD

**Issue by pcanilho** [jenkinsci/hashicorp-vault-plugin#201]
### Version report

Jenkins and plugins versions report:

```
Jenkins: 2.319.1
OS: Linux - 5.10.0-9-amd64
---
Office-365-Connector:4.15.2
ace-editor:1.1
ansible:1.1
ansicolor:1.0.1
ant:1.13
apache-httpcomponents-client-4-api:4.5.13-1.0
artifactory:3.13.2
atlassian-bitbucket-server-integration:3.1.0
authentication-tokens:1.4
basic-branch-build-strategies:1.3.2
blueocean:1.25.2
blueocean-autofavorite:1.2.4
blueocean-bitbucket-pipeline:1.25.2
blueocean-commons:1.25.2
blueocean-config:1.25.2
blueocean-core-js:1.25.2
blueocean-dashboard:1.25.2
blueocean-display-url:2.4.1
blueocean-events:1.25.2
blueocean-git-pipeline:1.25.2
blueocean-github-pipeline:1.25.2
blueocean-i18n:1.25.2
blueocean-jira:1.25.2
blueocean-jwt:1.25.2
blueocean-personalization:1.25.2
blueocean-pipeline-api-impl:1.25.2
blueocean-pipeline-editor:1.25.2
blueocean-pipeline-scm-api:1.25.2
blueocean-rest:1.25.2
blueocean-rest-impl:1.25.2
blueocean-web:1.25.2
bootstrap4-api:4.6.0-3
bootstrap5-api:5.1.3-3
bouncycastle-api:2.25
branch-api:2.7.0
build-timeout:1.20
caffeine-api:2.9.2-29.v717aac953ff3
checks-api:1.7.2
cloudbees-bitbucket-branch-source:734.v2f848c5e6ea2
cloudbees-disk-usage-simple:0.10
cloudbees-folder:6.16
cobertura:1.17
code-coverage-api:2.0.4
command-launcher:1.6
config-file-provider:3.8.2
configuration-as-code:1.55
credentials:2.6.2
credentials-binding:1.27
dashboard-view:2.18
data-tables-api:1.11.3-4
display-url-api:2.3.5
docker-commons:1.17
docker-java-api:3.1.5.2
docker-plugin:1.2.3
docker-workflow:1.26
durable-task:493.v195aefbb0ff2
echarts-api:5.2.2-1
email-ext:2.85
extended-read-permission:3.2
favorite:2.3.3
font-awesome-api:5.15.4-4
forensics-api:1.7.0
git:4.10.0
git-client:3.10.0
git-server:1.10
github:1.34.1
github-api:1.301-378.v9807bd746da5
github-branch-source:2.11.3
gradle:1.37.1
handlebars:3.0.8
handy-uri-templates-2-api:2.1.8-1.0
hashicorp-vault-pipeline:1.3
hashicorp-vault-plugin:336.v182c0fbaaeb7
htmlpublisher:1.28
ivy:2.1
jackson2-api:2.13.0-230.v59243c64b0a5
jacoco:3.3.0
javadoc:1.6
jaxb:2.3.0.1
jdk-tool:1.5
jenkins-design-language:1.25.2
jira:3.6
jira-steps:1.6.0
jjwt-api:0.11.2-9.c8b45b8bb173
jquery-detached:1.2.1
jquery3-api:3.6.0-2
jsch:0.1.55.2
junit:1.53
lockable-resources:2.12
mailer:1.34
matrix-auth:2.6.9
matrix-project:1.19
maven-plugin:3.15.1
metrics:4.0.2.8
momentjs:1.1.1
node-iterator-api:1.5.1
okhttp-api:4.9.3-105.vb96869f8ac3a
pipeline-build-step:2.15
pipeline-github-lib:1.0
pipeline-graph-analysis:1.12
pipeline-input-step:2.12
pipeline-milestone-step:1.3.2
pipeline-model-api:1.9.3
pipeline-model-definition:1.9.3
pipeline-model-extensions:1.9.3
pipeline-rest-api:2.19
pipeline-stage-step:2.5
pipeline-stage-tags-metadata:1.9.3
pipeline-stage-view:2.19
pipeline-utility-steps:2.11.0
plain-credentials:1.7
plugin-util-api:2.6.0
popper-api:1.16.1-2
popper2-api:2.10.2-1
prometheus:2.0.10
pubsub-light:1.16
resource-disposer:0.16
saml:2.0.9
scm-api:2.6.5
script-security:1.78
snakeyaml-api:1.29.1
sse-gateway:1.24
ssh-agent:1.23
ssh-credentials:1.19
ssh-slaves:1.33.0
sshd:3.1.0
structs:308.v852b473a2b8c
terraform:1.0.10
timestamper:1.15
token-macro:267.vcdaea6462991
trilead-api:1.0.13
variant:1.4
vsphere-cloud:2.26
workflow-aggregator:2.6
workflow-api:2.47
workflow-basic-steps:2.24
workflow-cps:2640.v00e79c8113de
workflow-cps-global-lib:548.v9085a486966a
workflow-durable-task-step:1101.vf832bc1ac745
workflow-job:2.42
workflow-multibranch:2.26
workflow-scm-step:2.13
workflow-step-api:2.24
workflow-support:3.8
ws-cleanup:0.39
```

- What Operating System are you using (both controller, and any agents involved in the problem)?

Our controller is only being used as a job scheduler, all jobs run on dynamically spawned workers, based on Debian 11 images templates, via the [vsphere-cloud](https://plugins.jenkins.io/vsphere-cloud/#releases) plugin. Therefore, this issue is only affecting the *workers*.

### Reproduction steps

- Install the latest Hashicorp Release `336.v182c0fbaaeb7` present here: [link](https://updates.jenkins.io/download/plugins/hashicorp-vault-plugin/336.v182c0fbaaeb7/hashicorp-vault-plugin.hpi).
- Run a standard pipeline that uses a SCM checkout step.

### Results

Expected result:

SCM checkout works.

Actual result:

```sh
java.io.NotSerializableException: hudson.model.Hudson
  	at java.base/java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1185)
  	at java.base/java.io.ObjectOutputStream.defaultWriteFields(ObjectOutputStream.java:1553)
  	at java.base/java.io.ObjectOutputStream.writeSerialData(ObjectOutputStream.java:1510)
  	at java.base/java.io.ObjectOutputStream.writeOrdinaryObject(ObjectOutputStream.java:1433)
  	at java.base/java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1179)
  	at java.base/java.io.ObjectOutputStream.writeArray(ObjectOutputStream.java:1379)
  	at java.base/java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1175)
  	at java.base/java.io.ObjectOutputStream.defaultWriteFields(ObjectOutputStream.java:1553)
  	at java.base/java.io.ObjectOutputStream.writeSerialData(ObjectOutputStream.java:1510)
  	at java.base/java.io.ObjectOutputStream.writeOrdinaryObject(ObjectOutputStream.java:1433)
  	at java.base/java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1179)
  	at java.base/java.io.ObjectOutputStream.writeObject(ObjectOutputStream.java:349)
  	at hudson.remoting.UserRequest._serialize(UserRequest.java:263)
  	at hudson.remoting.UserRequest.serialize(UserRequest.java:272)
  Caused: java.io.IOException: Unable to serialize UserRPCRequest:org.jenkinsci.plugins.gitclient.GitClient.addCredentials[java.lang.String,com.cloudbees.plugins.credentials.common.StandardCredentials](3)
  	at hudson.remoting.UserRequest.serialize(UserRequest.java:274)
  	at hudson.remoting.UserRequest.<init>(UserRequest.java:101)
  	at hudson.remoting.Channel.call(Channel.java:999)
  	at hudson.remoting.RemoteInvocationHandler.invoke(RemoteInvocationHandler.java:286)
  Caused: hudson.remoting.RemotingSystemException
  	at hudson.remoting.RemoteInvocationHandler.invoke(RemoteInvocationHandler.java:302)
  	at com.sun.proxy.$Proxy92.addCredentials(Unknown Source)
  	at org.jenkinsci.plugins.gitclient.RemoteGitImpl.addCredentials(RemoteGitImpl.java:201)
  	at hudson.plugins.git.GitSCM.createClient(GitSCM.java:920)
  	at hudson.plugins.git.GitSCM.createClient(GitSCM.java:838)
  	at hudson.plugins.git.GitSCM.checkout(GitSCM.java:1291)
  	at org.jenkinsci.plugins.workflow.steps.scm.SCMStep.checkout(SCMStep.java:129)
  	at org.jenkinsci.plugins.workflow.steps.scm.SCMStep$StepExecutionImpl.run(SCMStep.java:97)
  	at org.jenkinsci.plugins.workflow.steps.scm.SCMStep$StepExecutionImpl.run(SCMStep.java:84)
  	at org.jenkinsci.plugins.workflow.steps.SynchronousNonBlockingStepExecution.lambda$start$0(SynchronousNonBlockingStepExecution.java:47)
  	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
  	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
  	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
  	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
  	at java.base/java.lang.Thread.run(Thread.java:829)
```

Quite possibly related to the merge of October 1st: https://github.com/jenkinsci/hashicorp-vault-plugin/pull/178#discussion_r702989909

**Comment by UXabre** [comment]
I can confirm that downgrading to 3.8.0 works like a charm so must indeed be something broken in the last release. What I did notice was that this was working on my master node but not on any slave node (test both CentOS 7 and Windows Server 2019 slaves)

**Comment by jetersen** [comment]
I do not actively use Jenkins and Vault anymore.
So I would appreciate someone stepping up and willing to fix the issue.

cc @jamesrobson-secondmind

**Comment by jamesrobson-secondmind** [comment]
I can reproduce this using the last released version, but interestingly the version of the plugin I was using was built from [my own branch](https://github.com/jamesrobson-secondmind/hashicorp-vault-plugin/tree/add-gcp-service-accounts) while waiting for the previous release and that was not affected by this issue. That included #178 and #179 but not #195, possibly there's a conflict between my changes and that one.
I'll see if I can find the root cause.

**Comment by jamesrobson-secondmind** [comment]
My investigation points to #195 as the culprit, @hchakrouun could you help investigate this issue? 

To check this I rebuilt the current `master` from source then went back 1 commit at a time until the failure stopped happening. Commit 8c39ae40 had the failure and b53b6b6f did not. I then built the plugin from hchakrouun's repo to double check that was the cause and that also failed.
Could someone try building the plugin from hchakrouun's repo to confirm that my results are reproducible?

Here's what I ran:
```
git clone https://github.com/hchakrouun/hashicorp-vault-plugin.git hchakrouun-vault-plugin
cd hchakrouun-vault-plugin
mvn test && mvn hpi:hpi
```
then I ran a job with the following pipeline:
```
pipeline {
    agent {
        label 'linux'
} 
    stages {
        stage('1') {
            steps {
                git  branch: 'master', credentialsId: '<sshkey>', url: 'git@github.com:jenkinsci/hashicorp-vault-plugin.git'
                sh 'ls'
            }
        }
    }
}
```

I'll see if I can come up with a fix and a test to cover this, but reverting #195 may be the best short term fix to get a working plugin again.

**Comment by jetersen** [comment]
@jamesrobson-secondmind what if you put it as `transient` so it avoids serialization. It is only needed for execution purposes.

**Comment by jetersen** [comment]
Seems the correct usage is to create an provider:

https://github.com/jenkinsci/amazon-ecr-plugin/blob/master/src/main/java/com/cloudbees/jenkins/plugins/amazonecr/AmazonECSRegistryCredentialsProvider.java

**Comment by jamesrobson-secondmind** [comment]
Adding `transient` results in
```
Also:   hudson.remoting.Channel$CallSiteStackTrace: Remote call to channel
10:07:21  		at hudson.remoting.Channel.attachCallSiteStackTrace(Channel.java:1797)
10:07:21  		at hudson.remoting.UserRequest$ExceptionResponse.retrieve(UserRequest.java:356)
10:07:21  		at hudson.remoting.Channel.call(Channel.java:1001)
10:07:21  		at com.datapipe.jenkins.vault.credentials.common.VaultHelper.getVaultSecret(VaultHelper.java:48)
10:07:21  		at com.datapipe.jenkins.vault.credentials.common.VaultHelper.getVaultSecretKey(VaultHelper.java:64)
10:07:21  		at com.datapipe.jenkins.vault.credentials.common.AbstractVaultBaseStandardCredentials.getVaultSecretKeyValue(AbstractVaultBaseStandardCredentials.java:88)
10:07:21  		at com.datapipe.jenkins.vault.credentials.common.VaultSSHUserPrivateKeyImpl.getPrivateKey(VaultSSHUserPrivateKeyImpl.java:87)
10:07:21  		at com.datapipe.jenkins.vault.credentials.common.VaultSSHUserPrivateKeyImpl.getPrivateKeys(VaultSSHUserPrivateKeyImpl.java:93)
10:07:21  		at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.createSshKeyFile(CliGitAPIImpl.java:2289)
10:07:21  		at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.launchCommandWithCredentials(CliGitAPIImpl.java:2028)
10:07:21  		at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.access$500(CliGitAPIImpl.java:86)
10:07:21  		at org.jenkinsci.plugins.gitclient.CliGitAPIImpl$1.execute(CliGitAPIImpl.java:624)
10:07:21  		at org.jenkinsci.plugins.gitclient.CliGitAPIImpl$2.execute(CliGitAPIImpl.java:853)
10:07:21  		at org.jenkinsci.plugins.gitclient.RemoteGitImpl$CommandInvocationHandler$GitCommandMasterToSlaveCallable.call(RemoteGitImpl.java:160)
10:07:21  		at org.jenkinsci.plugins.gitclient.RemoteGitImpl$CommandInvocationHandler$GitCommandMasterToSlaveCallable.call(RemoteGitImpl.java:153)
10:07:21  		at hudson.remoting.UserRequest.perform(UserRequest.java:211)
10:07:21  		at hudson.remoting.UserRequest.perform(UserRequest.java:54)
10:07:21  		at hudson.remoting.Request$2.run(Request.java:376)
10:07:21  		at hudson.remoting.InterceptingExecutorService.lambda$wrap$0(InterceptingExecutorService.java:78)
10:07:21  java.lang.IllegalStateException: The vault credential need a context
10:07:21  	at com.datapipe.jenkins.vault.credentials.common.VaultHelper$SecretRetrieve.call(VaultHelper.java:104)
10:07:21  	at com.datapipe.jenkins.vault.credentials.common.VaultHelper$SecretRetrieve.call(VaultHelper.java:79)
10:07:21  	at hudson.remoting.UserRequest.perform(UserRequest.java:211)
10:07:21  	at hudson.remoting.UserRequest.perform(UserRequest.java:54)
10:07:21  	at hudson.remoting.Request$2.run(Request.java:376)
10:07:21  	at hudson.remoting.InterceptingExecutorService.lambda$wrap$0(InterceptingExecutorService.java:78)
10:07:21  	at org.jenkinsci.remoting.CallableDecorator.call(CallableDecorator.java:18)
10:07:21  	at hudson.remoting.CallableDecoratorList.lambda$applyDecorator$0(CallableDecoratorList.java:19)
10:07:21  	at jenkins.util.ContextResettingExecutorService$2.call(ContextResettingExecutorService.java:46)
10:07:21  	at jenkins.security.ImpersonatingExecutorService$2.call(ImpersonatingExecutorService.java:80)
10:07:21  Also:   hudson.remoting.Channel$CallSiteStackTrace: Remote call to cpu-standard-4biewt
10:07:21  		at hudson.remoting.Channel.attachCallSiteStackTrace(Channel.java:1797)
10:07:21  		at hudson.remoting.UserRequest$ExceptionResponse.retrieve(UserRequest.java:356)
10:07:21  		at hudson.remoting.Channel.call(Channel.java:1001)
10:07:21  		at org.jenkinsci.plugins.gitclient.RemoteGitImpl$CommandInvocationHandler.execute(RemoteGitImpl.java:145)
10:07:21  		at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
10:07:21  		at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
10:07:21  		at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
10:07:21  		at java.lang.reflect.Method.invoke(Method.java:498)
10:07:21  		at org.jenkinsci.plugins.gitclient.RemoteGitImpl$CommandInvocationHandler.invoke(RemoteGitImpl.java:131)
10:07:21  		at com.sun.proxy.$Proxy99.execute(Unknown Source)
10:07:21  		at hudson.plugins.git.GitSCM.retrieveChanges(GitSCM.java:1224)
10:07:21  		at hudson.plugins.git.GitSCM.checkout(GitSCM.java:1302)
10:07:21  		at org.jenkinsci.plugins.workflow.steps.scm.SCMStep.checkout(SCMStep.java:129)
10:07:21  		at org.jenkinsci.plugins.workflow.steps.scm.SCMStep$StepExecutionImpl.run(SCMStep.java:97)
10:07:21  		at org.jenkinsci.plugins.workflow.steps.scm.SCMStep$StepExecutionImpl.run(SCMStep.java:84)
10:07:21  		at org.jenkinsci.plugins.workflow.steps.SynchronousNonBlockingStepExecution.lambda$start$0(SynchronousNonBlockingStepExecution.java:47)
10:07:21  		at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
10:07:21  Caused: java.lang.IllegalStateException
10:07:21  	at com.datapipe.jenkins.vault.credentials.common.VaultHelper.getVaultSecretKey(VaultHelper.java:75)
10:07:21  	at com.datapipe.jenkins.vault.credentials.common.AbstractVaultBaseStandardCredentials.getVaultSecretKeyValue(AbstractVaultBaseStandardCredentials.java:88)
10:07:21  	at com.datapipe.jenkins.vault.credentials.common.VaultSSHUserPrivateKeyImpl.getPrivateKey(VaultSSHUserPrivateKeyImpl.java:87)
10:07:21  	at com.datapipe.jenkins.vault.credentials.common.VaultSSHUserPrivateKeyImpl.getPrivateKeys(VaultSSHUserPrivateKeyImpl.java:93)
10:07:21  	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.createSshKeyFile(CliGitAPIImpl.java:2289)
10:07:21  	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.launchCommandWithCredentials(CliGitAPIImpl.java:2028)
10:07:21  	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.access$500(CliGitAPIImpl.java:86)
10:07:21  	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl$1.execute(CliGitAPIImpl.java:624)
10:07:21  	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl$2.execute(CliGitAPIImpl.java:853)
10:07:21  	at org.jenkinsci.plugins.gitclient.RemoteGitImpl$CommandInvocationHandler$GitCommandMasterToSlaveCallable.call(RemoteGitImpl.java:160)
10:07:21  	at org.jenkinsci.plugins.gitclient.RemoteGitImpl$CommandInvocationHandler$GitCommandMasterToSlaveCallable.call(RemoteGitImpl.java:153)
10:07:21  	at hudson.remoting.UserRequest.perform(UserRequest.java:211)
10:07:21  	at hudson.remoting.UserRequest.perform(UserRequest.java:54)
10:07:21  	at hudson.remoting.Request$2.run(Request.java:376)
10:07:21  	at hudson.remoting.InterceptingExecutorService.lambda$wrap$0(InterceptingExecutorService.java:78)
10:07:21  	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
10:07:21  	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
10:07:21  	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
10:07:21  	at java.lang.Thread.run(Thread.java:748)
```

**Comment by jetersen** [comment]
@jamesrobson-secondmind I think the solution would be to use a string. Than serialize ItemGroup by doing `context.getFullName()` and use the [`Jenkins.getItemByFullName(String,Class)`](https://javadoc.jenkins.io/jenkins/model/Jenkins.html#getItemByFullName-java.lang.String-java.lang.Class-) to safely transfer the context over.

**Comment by jetersen** [comment]
Created #205 once a hpi file is ready, if someone is able to test it that would be great :)

**Comment by UXabre** [comment]
I'm willing to test this as soon as an hpi is available :-)

**Comment by jetersen** [comment]
HPI is available here: https://repo.jenkins-ci.org/incrementals/com/datapipe/jenkins/plugins/hashicorp-vault-plugin/343.vc11438a446ec/hashicorp-vault-plugin-343.vc11438a446ec.hpi
See the folder here:
https://repo.jenkins-ci.org/incrementals/com/datapipe/jenkins/plugins/hashicorp-vault-plugin/343.vc11438a446ec/

**Comment by UXabre** [comment]
So, running this, I get the same result unfortunately :-(

Jenkins v2.319.1
Hashicorp Vault Plugin v343.vc11438a446ec 

java.io.NotSerializableException: hudson.model.Hudson
	at java.base/java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1185)
	at java.base/java.io.ObjectOutputStream.defaultWriteFields(ObjectOutputStream.java:1553)
	at java.base/java.io.ObjectOutputStream.writeSerialData(ObjectOutputStream.java:1510)
	at java.base/java.io.ObjectOutputStream.writeOrdinaryObject(ObjectOutputStream.java:1433)
	at java.base/java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1179)
	at java.base/java.io.ObjectOutputStream.writeArray(ObjectOutputStream.java:1379)
	at java.base/java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1175)
	at java.base/java.io.ObjectOutputStream.defaultWriteFields(ObjectOutputStream.java:1553)
	at java.base/java.io.ObjectOutputStream.writeSerialData(ObjectOutputStream.java:1510)
	at java.base/java.io.ObjectOutputStream.writeOrdinaryObject(ObjectOutputStream.java:1433)
	at java.base/java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1179)
	at java.base/java.io.ObjectOutputStream.writeObject(ObjectOutputStream.java:349)
	at hudson.remoting.UserRequest._serialize(UserRequest.java:263)
	at hudson.remoting.UserRequest.serialize(UserRequest.java:272)
Caused: java.io.IOException: Unable to serialize UserRPCRequest:org.jenkinsci.plugins.gitclient.GitClient.addCredentials[java.lang.String,com.cloudbees.plugins.credentials.common.StandardCredentials](30)
	at hudson.remoting.UserRequest.serialize(UserRequest.java:274)
	at hudson.remoting.UserRequest.<init>(UserRequest.java:101)
	at hudson.remoting.Channel.call(Channel.java:999)
	at hudson.remoting.RemoteInvocationHandler.invoke(RemoteInvocationHandler.java:286)
Caused: hudson.remoting.RemotingSystemException
	at hudson.remoting.RemoteInvocationHandler.invoke(RemoteInvocationHandler.java:302)
	at com.sun.proxy.$Proxy613.addCredentials(Unknown Source)
	at org.jenkinsci.plugins.gitclient.RemoteGitImpl.addCredentials(RemoteGitImpl.java:201)
	at hudson.plugins.git.GitSCM.createClient(GitSCM.java:920)
	at hudson.plugins.git.GitSCM.createClient(GitSCM.java:838)
	at hudson.plugins.git.GitSCM.checkout(GitSCM.java:1291)
	at org.jenkinsci.plugins.workflow.steps.scm.SCMStep.checkout(SCMStep.java:129)
	at org.jenkinsci.plugins.workflow.steps.scm.SCMStep$StepExecutionImpl.run(SCMStep.java:97)
	at org.jenkinsci.plugins.workflow.steps.scm.SCMStep$StepExecutionImpl.run(SCMStep.java:84)
	at org.jenkinsci.plugins.workflow.steps.SynchronousNonBlockingStepExecution.lambda$start$0(SynchronousNonBlockingStepExecution.java:47)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829)

**Comment by hchakrouun** [comment]
Sorry guys,
It's been a while I didn't read my emails, so I didn't answer you @jamesrobson-secondmind 
I tried to reproduce the same thing with a simple pipeline:
```
pipeline {
    agent any

    stages {
        stage('1') {
            steps {
                 // This is a normal Jenkins credential: "64facc4f-41d0-4ae9-ab2a-43ceb1b67f60" and the git is cloning
                 // the repo using the SSH Key without any issue
                git  branch: 'master', credentialsId: '64facc4f-41d0-4ae9-ab2a-43ceb1b67f60', url: 'git@github.com:hchakrouun/hashicorp-vault-plugin.git'
                sh 'ls'
            }
        }

        stage('2') {
            environment {
                v = credentials('4d7fb51a-4c26-4afe-994d-9ef765ef5a93')
            }

            steps {
                 // This is for getting a SHH Key from Vault and I can get it without any problem
                script {
                    def jContent = readFile(file: v)
                    print jContent.collect { it }
                }

                 // But for **git** step to use the credential id of an SSH Key in Vault, this is not working,
                 // I am working in that to see why, but using credentials() and withCredentials(), it works and I tested it many times
                git  branch: 'master', credentialsId: '4d7fb51a-4c26-4afe-994d-9ef765ef5a93', url: 'git@github.com:hchakrouun/hashicorp-vault-plugin.git'
                sh 'ls'
            }
        }
    }
}
```

**Comment by jetersen** [comment]
see https://github.com/jenkinsci/azure-keyvault-plugin/pull/96#issuecomment-998830667 I tried to fix it in #206 but it has issues. I need to find time to reduce the changeset to only include snapshotter.

**Comment by hchakrouun** [comment]
I send a Pull Request early this morning to decode correctly the SSH Key https://github.com/jenkinsci/hashicorp-vault-plugin/pull/208 , with this modification both my git steps in the pipeline above worked.
This a command how I stored the SSH Key that I am using: 
`base64 privateKey | vault kv put kv/an/sshkey username="admin" passphrase="privatekey123" ssh_key=-`
privateKey: Is the SSH private Key file

**Comment by hchakrouun** [comment]
@jamesrobson-secondmind Do you still have the same error ?

**Comment by jamesrobson-secondmind** [comment]
I've checked with the latest code on #208, but it still has the same problem for me. Plus that change does nothing when using a `username+password` credential which also hits the serialisation exception.

Are you running the test on a remote agent? https://github.com/jenkinsci/hashicorp-vault-plugin/issues/201#issuecomment-994914623 mentions that jobs running on the master do not encounter this issue

**Comment by hchakrouun** [comment]
I'm using a local installation of Jenkins and as you can see, in my pipeline I used **any** and I didn't configure any remote agents, so Jenkins is picking any local agent to run my pipeline.
I'll try to configure one agent to reproduce the problem you are facing

**Comment by jetersen** [comment]
There are other reasons why #210 is not a valuable solution sadly 😓

**Comment by jetersen** [comment]
The best path forward is #206

**Comment by markush81** [comment]
> The best path forward is #206

was just trying out the #206 build, it results into

```
java.lang.NullPointerException
	at com.datapipe.jenkins.vault.credentials.common.VaultUsernamePasswordCredentialImpl.getUsername(VaultUsernamePasswordCredentialImpl.java:72)
	at com.datapipe.jenkins.vault.credentials.common.VaultUsernamePasswordCredentialBinding.bind(VaultUsernamePasswordCredentialBinding.java:53)
	at org.jenkinsci.plugins.credentialsbinding.impl.BindingStep$Execution2.doStart(BindingStep.java:134)
	at org.jenkinsci.plugins.workflow.steps.GeneralNonBlockingStepExecution.lambda$run$0(GeneralNonBlockingStepExecution.java:77)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829)
Finished: FAILURE
```

**Comment by jetersen** [comment]
@markush81 I did not say that #206 was finished 😅

Currently the build is failing so the build you picked up does not have a fix for the nullpointer 😅

**Comment by n0rthernstar** [comment]
Confirmed that issue exists when function is called 
checkout([$class: 'GitSCM',
branches: [[name: "${params.git_branch}"]],
doGenerateSubmoduleConfigurations: false,
extensions: [[$class: 'SubmoduleOption',
  disableSubmodules: false,
  parentCredentials: false,
  recursiveSubmodules: true,
  reference: '',
  trackingSubmodules: false]],
submoduleCfg: [],
userRemoteConfigs: [[credentialsId: 'secret-ir', url: 'SOME_GIT_URL']]
])
not on master but on jenkins slave node

**Comment by n0rthernstar** [comment]
Workaround: manual downgrade to https://updates.jenkins.io/download/plugins/hashicorp-vault-plugin/3.8.0/hashicorp-vault-plugin.hpi

**Comment by yamitsilv** [comment]
Same issue for us both for 336.v182c0fbaaeb7 and 3.8.0.
Also can't use 3.8.0 as need secret file support.

**Comment by jetersen** [comment]
If anyone is willing to test out, this HPI: https://repo.jenkins-ci.org/incrementals/com/datapipe/jenkins/plugins/hashicorp-vault-plugin/342.vea_cb_4ee8cdf8/

Hopefully it should resolve the issue.

**Comment by ericcitaire** [comment]
> If anyone is willing to test out, this HPI #218 : https://repo.jenkins-ci.org/incrementals/com/datapipe/jenkins/plugins/hashicorp-vault-plugin/342.vea_cb_4ee8cdf8/
> 
> Hopefully it should resolve the issue.

@jamesrobson-secondmind @jetersen 

I get this error :

```
java.lang.IllegalStateException: Jenkins.instance is missing. Read the documentation of Jenkins.getInstanceOrNull to see what you are doing wrong.
	at jenkins.model.Jenkins.get(Jenkins.java:808)
	at com.datapipe.jenkins.vault.credentials.common.VaultHelper.getVaultSecret(VaultHelper.java:36)
	at com.datapipe.jenkins.vault.credentials.common.VaultHelper.getVaultSecretKey(VaultHelper.java:97)
Also:   hudson.remoting.Channel$CallSiteStackTrace: Remote call to JNLP4-connect connection from 10.121.241.86/10.121.241.86:52820
		at hudson.remoting.Channel.attachCallSiteStackTrace(Channel.java:1784)
		at hudson.remoting.UserRequest$ExceptionResponse.retrieve(UserRequest.java:356)
		at hudson.remoting.Channel.call(Channel.java:1000)
		at org.jenkinsci.plugins.gitclient.RemoteGitImpl$CommandInvocationHandler.execute(RemoteGitImpl.java:143)
		at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
		at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
		at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
		at java.base/java.lang.reflect.Method.invoke(Method.java:566)
		at org.jenkinsci.plugins.gitclient.RemoteGitImpl$CommandInvocationHandler.invoke(RemoteGitImpl.java:129)
		at com.sun.proxy.$Proxy111.execute(Unknown Source)
		at hudson.plugins.git.GitSCM.retrieveChanges(GitSCM.java:1216)
		at hudson.plugins.git.GitSCM.checkout(GitSCM.java:1294)
		at org.jenkinsci.plugins.workflow.steps.scm.SCMStep.checkout(SCMStep.java:129)
		at org.jenkinsci.plugins.workflow.steps.scm.SCMStep$StepExecutionImpl.run(SCMStep.java:97)
		at org.jenkinsci.plugins.workflow.steps.scm.SCMStep$StepExecutionImpl.run(SCMStep.java:84)
		at org.jenkinsci.plugins.workflow.steps.SynchronousNonBlockingStepExecution.lambda$start$0(SynchronousNonBlockingStepExecution.java:47)
		at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
		at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
		at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
		at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
		at java.base/java.lang.Thread.run(Thread.java:829)
Caused: java.lang.IllegalStateException
	at com.datapipe.jenkins.vault.credentials.common.VaultHelper.getVaultSecretKey(VaultHelper.java:108)
	at com.datapipe.jenkins.vault.credentials.common.AbstractVaultBaseStandardCredentials.getVaultSecretKeyValue(AbstractVaultBaseStandardCredentials.java:77)
	at com.datapipe.jenkins.vault.credentials.common.VaultSSHUserPrivateKeyImpl.getPrivateKey(VaultSSHUserPrivateKeyImpl.java:106)
	at com.datapipe.jenkins.vault.credentials.common.VaultSSHUserPrivateKeyImpl.getPrivateKeys(VaultSSHUserPrivateKeyImpl.java:112)
	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.createSshKeyFile(CliGitAPIImpl.java:2279)
	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.launchCommandWithCredentials(CliGitAPIImpl.java:2022)
	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.access$500(CliGitAPIImpl.java:84)
	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl$1.execute(CliGitAPIImpl.java:618)
	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl$2.execute(CliGitAPIImpl.java:847)
	at org.jenkinsci.plugins.gitclient.RemoteGitImpl$CommandInvocationHandler$GitCommandMasterToSlaveCallable.call(RemoteGitImpl.java:158)
	at org.jenkinsci.plugins.gitclient.RemoteGitImpl$CommandInvocationHandler$GitCommandMasterToSlaveCallable.call(RemoteGitImpl.java:151)
	at hudson.remoting.UserRequest.perform(UserRequest.java:211)
	at hudson.remoting.UserRequest.perform(UserRequest.java:54)
	at hudson.remoting.Request$2.run(Request.java:376)
	at hudson.remoting.InterceptingExecutorService.lambda$wrap$0(InterceptingExecutorService.java:78)
	at java.base/java.util.concurrent.FutureTask.run(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
	at hudson.remoting.Engine$1.lambda$newThread$0(Engine.java:122)
	at java.base/java.lang.Thread.run(Unknown Source)
```

I guess this happening because the job is executed on an agent which is not the built-in agent.

**Comment by dhs-rec** [comment]
Yep, same here.

**Comment by jamesrobson-secondmind** [comment]
I have a potential fix if someone can test it out. You will need to update the ssh-credentials as well, so install the hpi files from https://repo.jenkins-ci.org/incrementals/com/datapipe/jenkins/plugins/hashicorp-vault-plugin/351.va_b_7013590975/ and https://repo.jenkins-ci.org/incrementals/org/jenkins-ci/plugins/ssh-credentials/278.v16ca_c8f9cb_d3/

**Comment by dhs-rec** [comment]
Looks good here.

**Comment by mifitous** [comment]
Hi @jamesrobson-secondmind : can you please confirm this will fix issue #215 as well?

**Comment by jamesrobson-secondmind** [comment]
> can you please confirm this will fix issue #215 as well?

I believe it will.

**Comment by jamesrobson-secondmind** [comment]
I have a set of updates below which should fix this issue, can someone test them and confirm they were able to do checkouts with ssh keys and with username+passwords?

**_IMPORTANT_**  the credentials plugin is only compatible with jenkins >= 2.340, so if you are using the LTS versions you will need  2.346.1 due to be released Wednesday, June 22. If you install the credentials plugin on a Jenkins instance below 2.340 it will fail to start up.

https://repo.jenkins-ci.org/incrementals/com/datapipe/jenkins/plugins/hashicorp-vault-plugin/351.va_b_7013590975/
https://repo.jenkins-ci.org/incrementals/org/jenkins-ci/plugins/ssh-credentials/279.vda_45d191b_b_a_b_/
https://repo.jenkins-ci.org/incrementals/org/jenkins-ci/plugins/credentials/1132.ve331b_e4fa_715/

**Comment by MatvonSki** [comment]
Hi,
**It's works.**

I tested: username+passwords

I have:
Jenkins 2.346.1 - install on RHEL8, systemd
plugins:
- Credentials Plugin -1132.ve331b_e4fa_715 
- SSH Credentials Plugin - 279.vda_45d191b_b_a_b_ 
- HashiCorp Vault Plugin - 351.va_b_7013590975 
credential
- credential type: "**Vault Username-Password Credential**" - credentialsId: vault_gitlab_user_passwd

and run this code:
. . .
        steps{
            script{
                git branch: "${envGitBranch}",
                    credentialsId: "${envGitCredentialsId}",
                    url: "${envGitUrl}"
                                
            }
        }

and: 
**Finished: SUCCESS**

[jenkins_hcv_test-log.txt](https://github.com/jenkinsci/hashicorp-vault-plugin/files/8967189/jenkins_hcv_test-log.txt)

--------------------------------------------------
earlier (when used old plugin - hcv:3.8.0):

The recommended git tool is: NONE
. . .
java.io.NotSerializableException: hudson.model.Hudson
. . .
Caused: java.io.IOException: Unable to serialize UserRPCRequest:org.jenkinsci.plugins.gitclient.GitClient.addCredentials[java.lang.String,com.cloudbees.plugins.credentials.common.StandardCredentials](13)
. . .

**Comment by MatvonSki** [comment]
Hi,
**It's works.**

I tested too: ssh keys
- credential type: "Vault SSH Username with private key Credential"; credential_id: vault_gitlab_user_ssh
. . .
            def envGitCredentialsId = 'vault_gitlab_user_ssh'
            def envGitUrl = 'git@gitlab.../....git'
. . .
`        steps{
            script{
                git branch: "${envGitBranch}",
                    credentialsId: "${envGitCredentialsId}",
                    url: "${envGitUrl}"
                                
            }
        }`

. . . build and end:

**Finished: SUCCESS**

**Comment by ghandim** [comment]
After updating Jenkins to 2.346.1 (latest LTS) and accidentally updating both Vault plugins to latest everything works as expected. 
![image](https://user-images.githubusercontent.com/14774649/177126669-29c33472-54c0-4c90-accf-af8c220d5c62.png)

**Comment by jetersen** [comment]
@jamesrobson-secondmind your PR still fixes a known security issue btw.

**Comment by jetersen** [comment]
So perhaps we create a new PR that removes the channel communication without all the snapshots ?

**Comment by jetersen** [comment]
We have a new simplified PR after updates to `ssh-credentials` and `credentials` plugin

Could I ask some of the participants here that are willing to test this change, thanks and much appreciated 🙇

Here is the HPI file available from the #228 PR: https://repo.jenkins-ci.org/incrementals/com/datapipe/jenkins/plugins/hashicorp-vault-plugin/343.v4fa_d08d7995b_/

**Comment by jetersen** [comment]
@cronik see
https://github.com/jenkinsci/hashicorp-vault-plugin/issues/201#issuecomment-1173800462

You haven't updated `credentials` nor `ssh-credentials`

**Comment by cronik** [comment]
I didn't realize those were prerequisites. After updating to the versions listed in the comment above the serialization did not occur. Thanks.

```
Jenkins: 2.359
---
credentials:1139.veb_9579fca_33b_
hashicorp-vault-plugin:351.vdb_f83a_1c6a_9d
ssh-credentials:291.v8211e4f8efb_c
```

**Comment by jglick** [comment]
So I suppose this issue can be closed?

**Comment by cronik** [comment]
I think I may have found another case where the serialization issue occurs when using vault credential types. This time with the `http_request` plugin. If a vault credential id is passed the followed serialization error occurs, but if a regular username password credential id is used it works fine.

```
java.io.NotSerializableException: hudson.model.Hudson
	at java.base/java.io.ObjectOutputStream.writeObject0(Unknown Source)
	at java.base/java.io.ObjectOutputStream.defaultWriteFields(Unknown Source)
	at java.base/java.io.ObjectOutputStream.writeSerialData(Unknown Source)
	at java.base/java.io.ObjectOutputStream.writeOrdinaryObject(Unknown Source)
	at java.base/java.io.ObjectOutputStream.writeObject0(Unknown Source)
	at java.base/java.io.ObjectOutputStream.defaultWriteFields(Unknown Source)
	at java.base/java.io.ObjectOutputStream.writeSerialData(Unknown Source)
	at java.base/java.io.ObjectOutputStream.writeOrdinaryObject(Unknown Source)
	at java.base/java.io.ObjectOutputStream.writeObject0(Unknown Source)
	at java.base/java.io.ObjectOutputStream.defaultWriteFields(Unknown Source)
	at java.base/java.io.ObjectOutputStream.writeSerialData(Unknown Source)
	at java.base/java.io.ObjectOutputStream.writeOrdinaryObject(Unknown Source)
	at java.base/java.io.ObjectOutputStream.writeObject0(Unknown Source)
	at java.base/java.io.ObjectOutputStream.writeObject(Unknown Source)
	at hudson.remoting.UserRequest._serialize(UserRequest.java:263)
	at hudson.remoting.UserRequest.serialize(UserRequest.java:272)
Caused: java.io.IOException: Unable to serialize jenkins.plugins.http_request.HttpRequestExecution@35824d38
	at hudson.remoting.UserRequest.serialize(UserRequest.java:274)
	at hudson.remoting.UserRequest.<init>(UserRequest.java:101)
	at hudson.remoting.Channel.call(Channel.java:998)
	at jenkins.plugins.http_request.HttpRequestStep$Execution.run(HttpRequestStep.java:404)
	at jenkins.plugins.http_request.HttpRequestStep$Execution.run(HttpRequestStep.java:383)
	at org.jenkinsci.plugins.workflow.steps.SynchronousNonBlockingStepExecution.lambda$start$0(SynchronousNonBlockingStepExecution.java:47)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Unknown Source)
	at java.base/java.util.concurrent.FutureTask.run(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
	at java.base/java.lang.Thread.run(Unknown Source)
```

Environment
```
Jenkins: 2.346.2
---
credentials 1129.vef26f5df883c
credentials-binding 523.vd859a_4b_122e6
hashicorp-vault-plugin 351.vdb_f83a_1c6a_9d
http_request 1.15
```

**Comment by jamesrobson-secondmind** [comment]
@cronik Can you check the version of `credentials` you are using, it needs to be [1139.veb_9579fca_33b_](https://github.com/jenkinsci/credentials-plugin/releases/tag/1139.veb_9579fca_33b_)

@jetersen How do we set the minimum version of credentials to the correct value? The ssh-credentials plugin gets a minimum version that's new enough, but for some reason the [listed minimum for credentials](https://plugins.jenkins.io/hashicorp-vault-plugin/#dependencies) is `2.6.1.1`

**Comment by jetersen** [comment]
We need to target a newer `jenkins.version` version to `2.319.3` and `bom` version to `2.319.x` needs to be bumped.

https://github.com/jenkinsci/hashicorp-vault-plugin/blob/master/pom.xml#L16

Hopefully `2.319.3` should do

As we would be getting these versions from 

https://github.com/jenkinsci/bom/blob/9076c8f94d496e10bf2bfe1c3f90023087a5f781/bom-2.332.x/pom.xml#L35-L44

If that is still to low, we need to target even newer `jenkins.version` version to `2.346.1` and `bom` version to `2.346.x`

**Comment by jetersen** [comment]
We could consider asking for a backport of https://github.com/jenkinsci/credentials-plugin/pull/327 to older jenkins versions but I am not sure that is worth it.

**Comment by cronik** [comment]
@jamesrobson-secondmind I listed the wrong version in comment above. It is `credentials 1139.veb_9579fca_33b_`. For some reason `jenkins-plugin-cli` reported a different version than what is in the UI, I'll have to check on that later.

**Comment by jglick** [comment]
File a bug for that plugin. https://github.com/jenkinsci/http-request-plugin/blob/d98ff4e40ac0521df4a094c7862e50b289c2496d/src/main/java/jenkins/plugins/http_request/auth/CredentialBasicAuthentication.java#L33 etc. is serialized to an agent, so it needs to call `snapshot` (or serialize username + password directly).

**Comment by cronik** [comment]
Created [JENKINS-69155](https://issues.jenkins.io/browse/JENKINS-69155) for http-request-plugin.
